tls/s2n_key_log.c

```diff
@@ -0,0 +1,147 @@
+        GUARD_AS_RESULT(s2n_stuffer_write_uint8(output, chars[lower]));
+    }
+
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_key_log_tls13_secret(struct s2n_connection *conn, struct s2n_blob *secret, s2n_secret_type_t secret_type)
+{
+    ENSURE_REF(conn);
+    ENSURE_REF(conn->config);
+    ENSURE_REF(secret);
+
+    /* only emit keys if the callback has been set */
+    if (!conn->config->key_log_cb) {
+        return S2N_RESULT_OK;
+    }
+
+    const uint8_t client_early_traffic_label[] = "CLIENT_EARLY_TRAFFIC_SECRET ";
+    const uint8_t client_handshake_label[] = "CLIENT_HANDSHAKE_TRAFFIC_SECRET ";
+    const uint8_t server_handshake_label[] = "SERVER_HANDSHAKE_TRAFFIC_SECRET ";
+    const uint8_t client_traffic_label[] = "CLIENT_TRAFFIC_SECRET_0 ";
+    const uint8_t server_traffic_label[] = "SERVER_TRAFFIC_SECRET_0 ";
+
+    const uint8_t *label = NULL;
+    uint8_t label_size = 0;
+
+    switch (secret_type) {
+        case S2N_CLIENT_EARLY_TRAFFIC_SECRET:
+            label = client_early_traffic_label;
+            label_size = sizeof(client_early_traffic_label) - 1;
+            break;
+        case S2N_CLIENT_HANDSHAKE_TRAFFIC_SECRET:
+            label = client_handshake_label;
+            label_size = sizeof(client_handshake_label) - 1;
+            break;
+        case S2N_SERVER_HANDSHAKE_TRAFFIC_SECRET:
+            label = server_handshake_label;
+            label_size = sizeof(server_handshake_label) - 1;
+            break;
+        case S2N_CLIENT_APPLICATION_TRAFFIC_SECRET:
+            label = client_traffic_label;
+            label_size = sizeof(client_traffic_label) - 1;
+            break;
+        case S2N_SERVER_APPLICATION_TRAFFIC_SECRET:
+            label = server_traffic_label;
+            label_size = sizeof(server_traffic_label) - 1;
+            break;
+        default:
+            BAIL(S2N_ERR_SAFETY);
+    }
+
+    const uint8_t len
+        = label_size
+        + S2N_TLS_RANDOM_DATA_LEN * HEX_ENCODING_SIZE
+        + 1 /* SPACE */
+        + secret->size * HEX_ENCODING_SIZE;
+
+    DEFER_CLEANUP(struct s2n_stuffer output, s2n_stuffer_free);
+    GUARD_AS_RESULT(s2n_stuffer_alloc(&output, len));
+
+    GUARD_AS_RESULT(s2n_stuffer_write_bytes(&output, label, label_size));
+    GUARD_RESULT(s2n_key_log_hex_encode(&output, conn->secure.client_random, S2N_TLS_RANDOM_DATA_LEN));
+    GUARD_AS_RESULT(s2n_stuffer_write_uint8(&output, ' '));
+    GUARD_RESULT(s2n_key_log_hex_encode(&output, secret->data, secret->size));
+
+    uint8_t *data = s2n_stuffer_raw_read(&output, len);
+    ENSURE_REF(data);
+
+    conn->config->key_log_cb(conn->config->key_log_ctx, conn, data, len);
+
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_key_log_tls12_secret(struct s2n_connection *conn)
+{
+    ENSURE_REF(conn);
+
+    /* only emit keys if the callback has been set */
+    if (!conn->config->key_log_cb) {
+        return S2N_RESULT_OK;
+    }
+
+    const uint8_t label[] = "CLIENT_RANDOM ";
```

Comment:
I know the spec you link says that we use "CLIENT_RANDOM" for the TLS1.2 master secret, but I'd say that's weird enough naming that it probably needs a comment.

Reply:
Also, what's the point of including the space? You exclude if from label_size, and you also don't write it bc you only write label_size bytes.

Reply:
> Also, what's the point of including the space? You exclude if from label_size, and you also don't write it bc you only write label_size bytes.

Inline strings are null terminated so i'm not trimming off the space but the null value.